aspnetcore/client-side/import-export-interop.md

```diff
@@ -0,0 +1,271 @@
+
+    https://github.com/dotnet/runtime/blob/release/7.0/src/mono/wasm/templates/templates/browser/main.js
+
+    For example, the runMainAndExit() is changing to dotnet.run().
+
+    This is tracked on the 7.0 doc tracking issue at https://github.com/dotnet/AspNetCore.Docs/issues/26364.
+    
+-->
+
+`main.js` of an app created from the `wasmbrowser` project template:
+
+```javascript
+import { dotnet } from './dotnet.js'
+
+const is_browser = typeof window != "undefined";
+if (!is_browser) throw new Error(`Expected to be running in a browser`);
+
+const { setModuleImports, getAssemblyExports, getConfig, runMainAndExit } = 
+  await dotnet
+    .withDiagnosticTracing(false)
+    .withApplicationArgumentsFromQuery()
+    .create();
+
+setModuleImports("main.js", {
+  window: {
+    location: {
+      href: () => globalThis.window.location.href
+    }
+  }
+});
+
+const config = getConfig();
+const exports = await getAssemblyExports(config.mainAssemblyName);
+const text = exports.MyClass.Greeting();
+console.log(text);
+
+document.getElementById("out").innerHTML = `${text}`;
+await runMainAndExit(config.mainAssemblyName, ["dotnet", "is", "great!"]);
+```
+
+To import a JS function so it can be called from C#, use the new `JSImportAttribute` on a matching method signature. The first parameter to the `JSImportAttribute` is the name of the JS function to import and the second parameter is the name of the module.
+
+In the following example, the `window.location.href` function is called from the `main.js` module when `GetHRef` method is called. In an app generated from either the `wasmbrowser` or `wasmconsole` templates, `GetHRef` is in `Program.cs` with the partial class namespace `MyClass`:
+
+```csharp
+[JSImport("window.location.href", "main.js")]
+internal static partial string GetHRef();
+```
+
+In the imported method signature, you can use .NET types for parameters and return values, which are marshalled automatically by the runtime. Use `JSMarshalAsAttribute<T>` to control how the imported method parameters are marshalled. For example, you might choose to marshal a `long` as <xref:System.Runtime.InteropServices.JavaScript.JSType.Number?displayProperty=nameWithType> or <xref:System.Runtime.InteropServices.JavaScript.JSType.BigInt?displayProperty=nameWithType>. You can pass <xref:System.Action>/<xref:System.Func%601> callbacks as parameters, which are marshalled as callable JS functions. You can pass both JS and managed object references, and they are marshaled as proxy objects, keeping the object alive across the boundary until the proxy is garbage collected. You can also import and export asynchronous methods with a <xref:System.Threading.Tasks.Task> result, which are marshaled as [JS promises](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise). Most of the marshalled types work in both directions, as parameters and as return values, on both imported and exported methods.
```

Comment:
Here a table of supported mappings

| .NET                            | JavaScript | Nullable      | Task -> Promise | JSType.Missing | Array of |
|---------------------------------|------------|---------------|-----------------|----------------|----------|
| Boolean                         | Boolean    | ✅             | ✅               | ✅              |          |
| Byte                            | Number     | ✅             | ✅               | ✅              | ✅        |
| Char                            | String     | ✅             | ✅               | ✅              |          |
| Int16                           | Number     | ✅             | ✅               | ✅              |          |
| Int32                           | Number     | ✅             | ✅               | ✅              | ✅        |
| Int64                           | Number     | ✅             | ✅               |                |          |
| Int64                           | BigInt     | ✅             | ✅               |                |          |
| Single                          | Number     | ✅             | ✅               | ✅              |          |
| Double                          | Number     | ✅             | ✅               | ✅              | ✅        |
| IntPtr                          | Number     | ✅             | ✅               | ✅              |          |
| DateTime                        | Date       | ✅             | ✅               |                |          |
| DateTimeOffset                  | Date       | ✅             | ✅               |                | ✅        |
| Exception                       | Error      |               |                 | ✅              |          |
| JSObject                        | Object     |               |                 | ✅              | ✅        |
| String                          | String     |               |                 | ✅              |          |
| Object                          | Any        |               | ✅               |                | ✅        |
| Span&lt;Byte&gt;                | MemoryView |               |                 |                |          |
| Span&lt;Int32&gt;               | MemoryView |               |                 |                |          |
| Span&lt;Double&gt;              | MemoryView |               |                 |                |          |
| ArraySegment&lt;Byte&gt;        | MemoryView |               |                 |                |          |
| ArraySegment&lt;Int32&gt;       | MemoryView |               |                 |                |          |
| ArraySegment&lt;Double&gt;      | MemoryView |               |                 |                |          |
| Task                            | Promise    |               |                 |                |          |
| Action                          | Function   |               |                 |                |          |
| Action&lt;T1&gt;                | Function   |               |                 |                |          |
| Action&lt;T1, T2&gt;            | Function   |               |                 |                |          |
| Action&lt;T1, T2, T3&gt;        | Function   |               |                 |                |          |
| Func&lt;TResult&gt;             | Function   |               |                 |                |          |
| Func&lt;T1, TResult&gt;         | Function   |               |                 |                |          |
| Func&lt;T1, T2, TResult&gt;     | Function   |               |                 |                |          |
| Func&lt;T1, T2, T3, TResult&gt; | Function   |               |                 |                |          |

Reply:
@danroth27 I told you I'm not the only one who knows about this stuff :-D

Few more notes
 - `JSType.Missing` means that the `JSMarshalAs<>` annotation is optional. 
     - @guardrex you can figure out better name for that column.
     - I think it's also optional for void `Task`, there shuld be checkmark too.
 - The `Array of` is C# `string[]` mapped to JS Array full of strings. 
     - It will copy the array and marshal the values.
     - `DateTimeOffset` and `JSObject` should not have check 
     - `String` should have a check
 - when you are passing JS value to C# side and you give value of wrong type, it will throw in most cases. When running, we don't have compile time type checker in JS.
 - `JSObject`, `Exception`, `Task` and `ArraySegment` create GCHandle and a proxy. Could be disposed by user or GC would handle it later. They are not cheap.

@maraf could you please edit the grid ? Somehow I could not do it.

Reply:
`MemoryView` is a type which dotnet brings, not a native JS type. It's API is described in `dotnet.d.ts`
It doesn't copy the bytes during marshaling.
`MemoryView` created for `Span` is only valid for the duration of the interop call.
`MemoryView` created for `ArraySegment` is good for long term sharing of a buffer. It will survive after the interop call.
You can't create instances of `MemoryView` on JS side yourself. So marshaling them in this direction doesn't make sense.

Reply:
Can I "dagger" (&dagger;) the "`JSType.Missing`" column name (`JSType.Missing`&dagger;); and then under the table, dagger the remark like this ...

&dagger;`JSType.Missing` means that the `JSMarshalAs<T>` annotation is optional.

If that's ok, it keeps the column ***narrow***. On small device screens, we have trouble displaying wide tables. It's best to keep them as narrow as possible.

Reply:
I have updated the table.

> `DateTimeOffset` and `JSObject` should not have check

The `DateTimeOffset` was a mistake, but `JSObject` is there https://github.com/maraf/runtime/blob/99d21b9276ebe8f7bea7fb3ba74dca9fca625fe2/src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSGeneratorFactory.cs#L213

Reply:
`Task<string>` 
`Task<Exception>`
`Task<JSObject>`
`Task<object>` are all supported and working

Reply:
Also for `Task<string>`  `Task<Exception>` `Task<JSObject>` the annotation is optional.

Reply:
Yes, already marked in the table.

Reply:
@guardrex maybe we want to talk about "JSMarshalAs optional" rather as "these cases require JSMarshalAs annotation". I guess people would not annotate (and care about it) unless it's required.

Reply:
> `MemoryView` is a type which dotnet brings, not a native JS type. It's API is described in `dotnet.d.ts` It doesn't copy the bytes during marshaling. `MemoryView` created for `Span` is only valid for the duration of the interop call. `MemoryView` created for `ArraySegment` is good for long term sharing of a buffer. It will survive after the interop call. You can't create instances of `MemoryView` on JS side yourself. So marshaling them back to C# doesn't make sense.

Does it mean that:

1. It is not practical (or not possible) to export a .NET method that returns `Span`?
2. For exported method that returns `ArraySegment`, is it important to call `dispose()` in the returned `MemoryView` in `try... finally`?

Reply:
@FranklinWhale 
1) it's not possible. The Span is allocated on stack and it has GC implications. When calling from JS to C#, there is no C# stack after the call.
2) calling `dispose()` disposes the proxy and unpins the underlying C# byte array. It is good idea to do so. If you don't do it, the JS GC will do it eventually.
3) you can also return `byte[]` which will copy the bytes, but it's easiest to think about it.